src/MeteePaymaster.sol

```diff
@@ -0,0 +1,103 @@
+        verifyingSigner = _verifyingSigner;
+        metaPaymaster = _metaPaymaster;
+    }
+
+    /**
+     * return the hash we're going to sign off-chain (and validate on-chain)
+     * this method is called by the off-chain service, to sign the request.
+     * it is called on-chain from the validatePaymasterUserOp, to validate the signature.
+     * note that this signature covers all fields of the UserOperation, except the "paymasterAndData",
+     * which will carry the signature itself.
+     */
+    function getHash(UserOperation calldata userOp, uint48 validUntil, uint48 validAfter)
+    public view returns (bytes32) {
+        // can't use userOp.hash(), since it contains also the paymasterAndData itself.
+        return keccak256(
+            abi.encode(
+                userOp.getSender(),
+                userOp.nonce,
+                calldataKeccak(userOp.initCode),
+                calldataKeccak(userOp.callData),
+                userOp.callGasLimit,
+                userOp.verificationGasLimit,
+                userOp.preVerificationGas,
+                userOp.maxFeePerGas,
+                userOp.maxPriorityFeePerGas,
+                block.chainid,
+                address(this),
+                validUntil,
+                validAfter
+            )
+        );
+    }
+
+    /**
+     * verify our external signer signed this request.
+     * the "paymasterAndData" is expected to be the paymaster and a signature over the entire request params
+     * paymasterAndData[:20] : address(this)
+     * paymasterAndData[20:84] : abi.encode(validUntil, validAfter)
+     * paymasterAndData[84:] : signature
+     */
+    function _validatePaymasterUserOp(UserOperation calldata userOp, bytes32 /*userOpHash*/, uint256 /*requiredPreFund*/)
+    internal override view returns (bytes memory context, uint256 validationData) {
+        (uint48 validUntil, uint48 validAfter, bytes calldata signature) = parsePaymasterAndData(userOp.paymasterAndData);
+        // Only support 65-byte signatures, to avoid potential replay attacks.
+        require(signature.length == 65, "Paymaster: invalid signature length in paymasterAndData");
+        bytes32 hash = ECDSA.toEthSignedMessageHash(getHash(userOp, validUntil, validAfter));
+
+        // don't revert on signature failure: return SIG_VALIDATION_FAILED
+        if (verifyingSigner != ECDSA.recover(hash, signature)) {
+            return ("", _packValidationData(true, validUntil, validAfter));
+        }
+
+        // no need for other on-chain validation: entire UserOp should have been checked
+        // by the external service prior to signing it.
+        return (" ", _packValidationData(false, validUntil, validAfter));
+    }
+
+    function _postOp(PostOpMode mode, bytes calldata /*context*/, uint256 actualGasCost) internal override {
+        if (mode != PostOpMode.postOpReverted) {
```

Comment:
would we want this to happen for both success and user op reverted cases? https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/interfaces/IPaymaster.sol#L11 

would it be better to do == instead of != in case other PostOpMode values get added? not a huge deal since the contract is not upgradeable so it's unlikely we'd run into issues here, but would be more explicit

Reply:
the paymaster still pays for both the success and revert cases... we just don't want to call it twice (for example, if `metaPaymaster.fund` reverts, there's no need to call it twice)